## Homework 5

### Сингулярное разложение матриц

**1. Найти с помощью NumPy SVD для матрицы**

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}$$

In [1]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

In [4]:
A = np.array([[1, 2, 0],
              [0, 0, 5],
              [3, -4, 2],
              [1, 6, 5],
              [0, 1, 0]])
print(f'Матрица A:\n{A}')

Матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [5]:
U, s, VT = np.linalg.svd(A)

# s - список диагональных элементов, его нужно привести к виду диагональной матрицы для наглядности
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

In [6]:
print(f'Матрица U:\n{U}')
print(f'Матрица D:\n{D}')
print(f'Матрица VT:\n{V}')

Матрица U:
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]
Матрица D:
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]
Матрица VT:
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]


In [8]:
# проверяем
print(f'Матрица A:\n{A}')

print(f'Матрица U*D*VT:\n{np.dot(np.dot(U, D), VT)}')

Матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]
Матрица U*D*VT:
[[ 1.  2.  0.]
 [ 0. -0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [-0.  1.  0.]]


**2. Для матрицы из предыдущего задания найти:**

а) евклидову норму;

Евклидова норма матрицы равна евклидовой норме диагональной матрицы из ее сингулярных чисел $D$:

$$\left \| A \right \|_{E}=\mu_{1} = 8.82$$

б) норму Фробениуса.

In [10]:
np.linalg.norm(A, ord = 'fro')

11.045361017187261